agent/agent.go

```diff
@@ -262,7 +262,9 @@ func (a *agent) createTailnet(ctx context.Context, derpMap *tailcfg.DERPMap) (ne
 		return nil, xerrors.Errorf("create tailnet: %w", err)
 	}
 	defer func() {
-		if err != nil {
+		// Apparently, it's possible to shut down network via agent.Close().
+		// The condition `network != nil` prevents panic in tests.
+		if err != nil && network != nil {
```

Comment:
This one is actually a bit sneaky. This doesn't capture the original intent. We should probably rename either the return variable or the variable in `network, err =`. We want to do close on the function local network which is always non-nil here, not on the return variable which is nil if `return nil, err`.

Reply:
> We want to do close on the function local network which is always non-nil here, not on the return variable which is nil if return nil, err.

Looking at the source of `createTailnet`, it doesn't seem to be possible to have a nilled network, unless ...

`a.network = network` is set in 2 places: [first](https://github.com/coder/coder/blob/main/agent/agent.go#L218) and [second](https://github.com/coder/coder/blob/061635c36d6d73f27421e361a01e0976ecd8477e/agent/agent.go#L269). In the first, it's guarded with a mutex. 

It seems natural to reduce one `a.network = network`? I guess that's the only option to have a `network = nil`.

Reply:
Ok, we had an offline chat with Mathias and came up with a bit different solution.